### ISSUE TITLE
Crashes during macro expansion

### ISSUE THREAD

**Issue by MikeInnes** [JuliaLang/julia#8707]
I'm working on a macro, [here](https://github.com/one-more-minute/ObjectiveC.jl/blob/8f3541e1a9c1251f97a9611c2ba6095efe2395fe/src/syntax.jl), which seems to cause problems for the compiler.

This expands and evaluates just fine:

``` jlcon
julia> @objc [[[NSString alloc] initWithUTF8String:"foo"] UTF8String] |> bytestring
"foo"
```

But this hangs at the REPL:

``` jlcon
julia> :(@objc begin
           s = [NSString alloc]
           [s initWithUTF8String:"foo"]
       end) |> macroexpand
fatal: error thrown and no exception handler available.
Base.MethodError(f=getindex, args=(<?::quote Expr(:macrocall, :@objc, Expr(:block, Expr(:line, 2, :/Users/mike/Library/Application Support/LightTable/plugins/Julia-LT/jl/init.jl)::Any, :s = Expr(:hcat, :NSString, :alloc)::Any, # line 3, Expr(:hcat, :s, Expr(::, :initWithUTF8String, "foo")::Any)::Any)::Any)::Any end>, 1))
```

Julia seems to produce this error + occasional type errors and segfaults.

The weirdest part is that all is fine if you get rid of the call to `esc` [here](https://github.com/one-more-minute/ObjectiveC.jl/blob/8f3541e1a9c1251f97a9611c2ba6095efe2395fe/src/syntax.jl#L21).

**Comment by jakebolewski** [comment]
Please reopen if this is still an issue.

**Comment by one-more-minute** [comment]
Yup. Although no I get a GC segfault every time as opposed to a compiler issue.

Easiest way to reproduce is to check out [this branch](https://github.com/one-more-minute/ObjectiveC.jl/tree/julia-segfault), build, and enter:
```julia
julia> :(@objc begin
                  s = [NSString alloc]
                  [s initWithUTF8String:"foo"]
              end) |> macroexpand

signal (11): Segmentation fault: 11
__pool_alloc at /Users/mike/julia/src/gc.c:933
newstruct at /Users/mike/julia/src/./julia_internal.h:30
jl_deserialize_value_ at /Users/mike/julia/src/dump.c:1294
jl_deserialize_value_ at /Users/mike/julia/src/dump.c:1105
jl_deserialize_value_ at /Users/mike/julia/src/dump.c:1105
jl_deserialize_value_ at /Users/mike/julia/src/dump.c:1105
jl_deserialize_value_ at /Users/mike/julia/src/dump.c:1105
jl_uncompress_ast at /Users/mike/julia/src/dump.c:1620
jl_prepare_ast at /Users/mike/julia/src/ast.c:942
typeinf_uncached at ./inference.jl:1483
jlcall_typeinf_uncached_140453 at /Users/mike/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
typeinf at ./inference.jl:1403
jlcall_typeinf_140468 at /Users/mike/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
typeinf_ext at ./inference.jl:1347
jl_apply at /Users/mike/julia/src/./julia.h:1261
jl_apply at /Users/mike/julia/src/./julia.h:668
jl_mt_assoc_by_type at /Users/mike/julia/src/gf.c:1069
jl_apply_generic at /Users/mike/julia/src/gf.c:1694
print_response at REPL.jl:133
jlcall_print_response_158156 at  (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
print_response at REPL.jl:118
jlcall_print_response_158155 at  (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
anonymous at REPL.jl:595
run_interface at ./LineEdit.jl:1552
jlcall_run_interface_155191 at /Users/mike/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
run_frontend at ./REPL.jl:835
run_repl at ./REPL.jl:164
jl_apply at /Users/mike/julia/src/./julia.h:1261
_start at ./client.jl:444
jlcall__start_155894 at /Users/mike/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
true_main at /usr/local/bin/julia (unknown line)
main at /usr/local/bin/julia (unknown line)
Segmentation fault: 11
Mikes-MBP:~ mike$ 
```

**Comment by one-more-minute** [comment]
To narrow it down as far as I have time for, this is sufficient to reproduce the segfault on 0.4:

```julia
using Lazy

function calltransform(ex::Expr)
  obj = objcm(ex.args[1])
  args = ex.args[2:end]
  isempty(args) && error()
  if isexpr(args[1], Symbol)
    length(args) > 1 && error()
    return :(message($obj, Selector(args[1])))
  end
  all(arg->isexpr(arg, :(:)) && isexpr(arg.args[1], Symbol), args) || error()
  msg = join(vcat([arg.args[1] for arg in args], ""), ":")
  args = [objcm(arg.args[2]) for arg in args]
  :(message($obj, $msg, $(args...)))
end

objcm(ex::Expr) =
  isexpr(ex, :hcat) ? calltransform(ex) :
  Expr(ex.head, map(objcm, ex.args)...)

objcm(ex) = esc(ex)

macro objc(ex)
  objcm(ex)
end

:(@objc begin
    s = [NSString alloc]
    [s initWithUTF8String:"foo"]
end) |> macroexpand
```

On the latest 0.3 it just hangs indefinitely.

**Comment by one-more-minute** [comment]
Just got this but can't reproduce it:
```julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+4133 (2015-04-02 05:38 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 09ca2e1* (0 days old master)
|__/                   |  x86_64-apple-darwin14.1.0

julia> using ObjectiveC

signal (11): Segmentation fault: 11
__pool_alloc at /Users/mike/julia/src/gc.c:933
newstruct at /Users/mike/julia/src/./julia_internal.h:30
jl_cellset at /Users/mike/julia/src/./julia.h:685
jl_cellset at /Users/mike/julia/src/./julia.h:685
jl_cellset at /Users/mike/julia/src/./julia.h:685
jl_cellset at /Users/mike/julia/src/./julia.h:685
jl_cellset at /Users/mike/julia/src/./julia.h:685
jl_eh_restore_state at /Users/mike/julia/src/./julia.h:1349
jl_parse_eval_all at /Users/mike/julia/src/toplevel.c:539
jl_load at /Users/mike/julia/src/toplevel.c:592
include at ./boot.jl:250
jl_apply at /Users/mike/julia/src/./julia.h:1261
include_from_node1 at ./loading.jl:129
jl_apply at /Users/mike/julia/src/./julia.h:1261
jl_apply at /Users/mike/julia/src/interpreter.c:53
eval at /Users/mike/julia/src/interpreter.c:214
jl_toplevel_eval_flex at /Users/mike/julia/src/toplevel.c:502
jl_eval_module_expr at /Users/mike/julia/src/toplevel.c:148
jl_toplevel_eval_flex at /Users/mike/julia/src/toplevel.c:394
jl_parse_eval_all at /Users/mike/julia/src/toplevel.c:550
jl_load at /Users/mike/julia/src/toplevel.c:592
include at ./boot.jl:250
jl_apply at /Users/mike/julia/src/./julia.h:1261
include_from_node1 at ./loading.jl:129
jl_apply at /Users/mike/julia/src/./julia.h:1261
jl_apply at /Users/mike/julia/src/interpreter.c:53
eval at /Users/mike/julia/src/interpreter.c:214
jl_toplevel_eval_flex at /Users/mike/julia/src/toplevel.c:502
jl_eh_restore_state at /Users/mike/julia/src/./julia.h:1349
reload_path at ./loading.jl:153
_require at ./loading.jl:68
jl_apply at /Users/mike/julia/src/./julia.h:1261
require at ./loading.jl:51
jlcall_require_179345 at /Users/mike/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/mike/julia/src/./julia.h:1261
jl_apply at /Users/mike/julia/src/toplevel.c:338
eval_import_path at /Users/mike/julia/src/toplevel.c:364
jl_eh_restore_state at /Users/mike/julia/src/./julia.h:1349
eval_user_input at REPL.jl:60
jlcall_eval_user_input_180835 at  (unknown line)
jl_apply at /Users/mike/julia/src/gf.c:1693
anonymous at task.jl:89
jl_apply at /Users/mike/julia/src/task.c:233
Segmentation fault: 11
```

**Comment by simonbyrne** [comment]
This seems to be fixed in 0.5:
```
ulia> using MacroTools, ObjectiveC

WARNING: deprecated syntax "[a=>b for (a,b) in c]".
Use "Dict(a=>b for (a,b) in c)" instead.
ERROR: LoadError: LoadError: syntax: space before "(" not allowed in "classes ("
 in include_from_node1(::String) at ./loading.jl:488
 in include_from_node1(::String) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 in include_from_node1(::String) at ./loading.jl:488
 in include_from_node1(::String) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 in eval(::Module, ::Any) at ./boot.jl:234
 in eval(::Module, ::Any) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
 in require(::Symbol) at ./loading.jl:415
 in require(::Symbol) at /Applications/Julia-0.5.app/Contents/Resources/julia/lib/julia/sys.dylib:?
while loading /Users/simon/.julia/v0.5/ObjectiveC/src/syntax.jl, in expression starting on line 42
while loading /Users/simon/.julia/v0.5/ObjectiveC/src/ObjectiveC.jl, in expression starting on line 20

julia> using MacroTools, ObjectiveC

julia> @expand @objc begin
                  bundle = [NSBundle bundleWithPath:path]
                  bundle.ptr |> Int |> Bool || error("Bundle $path not found")
                  [bundle load] |> Bool || error("Couldn't load bundle $path")
                  return
                end
quote 
    $(Expr(:escape, :( # REPL[2], line 2:)))
    $(Expr(:escape, :(bundle = [NSBundle bundleWithPath:path])))
    $(Expr(:escape, :( # REPL[2], line 3:)))
    $(Expr(:escape, :((bundle.ptr |> Int) |> Bool || error("Bundle $(path) not found"))))
    $(Expr(:escape, :( # REPL[2], line 4:)))
    $(Expr(:escape, :([bundle load] |> Bool || error("Couldn't load bundle $(path)"))))
    $(Expr(:escape, :( # REPL[2], line 5:)))
    $(Expr(:escape, :(return)))
end
```
